# Clean Transform MBS Subsidised Services & Benefits Data

The script loads MBS datasets 2013-19, 2019-21 and 2021-22 and transforms each set to have measure values as separate columns instead of rows. The each dataset is exported to csv file. 

In [69]:
# import libraries
import pandas as pd
import numpy as np
import os

## 1.MBS 2021-2022 Data Cleaning & Tranformation

In [70]:
# import the mbs data files and assign to a dataframe

# setup path to original dataset of mbs fileds
path = r'/Users/patel/Documents/CF-Data Anaylst Course/portfolio_projects/mbs_analysis/datasets/'

#import required columns
filter_cols = ["Worksheet","Year","StateTerritory","GeographicUnit","GeographicCode","GeographicAreaName","GeographicGroup", "ServiceLevel", "Service", "DemographicGroup","MeasureName","MeasureValue","Flag","FlagExplanation"]

df_mbs_2021 = pd.read_csv(os.path.join(path, 'original_datasets/mbs_data/phc-mbs-2021-2022/AIHW_MBS_2122_CSV.csv'), encoding='ISO-8859-1', usecols=filter_cols, index_col=None)
df_mbs_2021.head(3)

/var/folders/6w/mdyngzzd2_z7brmk404qky_m0000gs/T/ipykernel_70532/485236433.py:9: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  df_mbs_2021 = pd.read_csv(os.path.join(path, 'original_datasets/mbs_data/phc-mbs-2021-2022/AIHW_MBS_2122_CSV.csv'), encoding='ISO-8859-1', usecols=filter_cols, index_col=None)


Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
0       PHN  2021-22       National            NAT         001NAT   
1       PHN  2021-22       National            NAT         001NAT   
2       PHN  2021-22       National            NAT         001NAT   

  GeographicAreaName GeographicGroup ServiceLevel                 Service  \
0           National        National      Level 1  GP attendances (total)   
1           National        National      Level 1  GP attendances (total)   
2           National        National      Level 1  GP attendances (total)   

  DemographicGroup                                   MeasureName  \
0             0-14  Percentage of people who had the service (%)   
1             0-14                       Services per 100 people   
2             0-14              Medicare benefits per 100 people   

   MeasureValue Flag FlagExplanation  
0       84.14%   NaN             NaN  
1        395.16  NaN             NaN  
2     17,792.31  NaN             NaN

In [71]:
df_mbs_2021.shape

(364176, 14)

In [72]:
df_mbs_2021.head()

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
0       PHN  2021-22       National            NAT         001NAT   
1       PHN  2021-22       National            NAT         001NAT   
2       PHN  2021-22       National            NAT         001NAT   
3       PHN  2021-22       National            NAT         001NAT   
4       PHN  2021-22       National            NAT         001NAT   

  GeographicAreaName GeographicGroup ServiceLevel                 Service  \
0           National        National      Level 1  GP attendances (total)   
1           National        National      Level 1  GP attendances (total)   
2           National        National      Level 1  GP attendances (total)   
3           National        National      Level 1  GP attendances (total)   
4           National        National      Level 1  GP attendances (total)   

  DemographicGroup                                   MeasureName  \
0             0-14  Percentage of people who had the service (%)   
1             0-14                       Services per 100 people   
2             0-14              Medicare benefits per 100 people   
3             0-14                               No. of patients   
4             0-14                               No. of services   

   MeasureValue Flag FlagExplanation  
0       84.14%   NaN             NaN  
1        395.16  NaN             NaN  
2     17,792.31  NaN             NaN  
3     3,997,279  NaN             NaN  
4    18,772,343  NaN             NaN

### 1.2 Extract and transform SA3 Worksheet Data

SA3 worksheet contains Medicare subsidsed servicees data by Statistical Area Level 3 (geograpghical area) 

In [73]:
# check the worksheet names to find SA3 reference
df_mbs_2021['Worksheet'].unique()

array(['PHN', 'SA3', 'PHN - Age standardised', 'PHN - GP aged care'],
      dtype=object)

In [74]:
df_mbs_2021[df_mbs_2021['Worksheet']=='SA3'].count()

Worksheet             231712
Year                  231712
StateTerritory        227680
GeographicUnit        231712
GeographicCode        231712
GeographicAreaName    231712
GeographicGroup       231712
ServiceLevel          231712
Service               231712
DemographicGroup      231712
MeasureName           231712
MeasureValue          231712
Flag                   16607
FlagExplanation        16607
dtype: int64

In [75]:
# extract SA3 data into a new dataframe. Used the copy function so the original dataframe is not impacted.
df_mbs_2021_sa3 = df_mbs_2021.loc[df_mbs_2021['Worksheet']=='SA3'].copy()
df_mbs_2021_sa3.head()

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
132192       SA3  2021-22       National            NAT         001NAT   
132193       SA3  2021-22       National            NAT         001NAT   
132194       SA3  2021-22       National            NAT         001NAT   
132195       SA3  2021-22       National            NAT         001NAT   
132196       SA3  2021-22       National            NAT         001NAT   

       GeographicAreaName GeographicGroup ServiceLevel  \
132192           National        National      Level 1   
132193           National        National      Level 1   
132194           National        National      Level 1   
132195           National        National      Level 1   
132196           National        National      Level 1   

                       Service DemographicGroup  \
132192  GP attendances (total)             0-24   
132193  GP attendances (total)             0-24   
132194  GP attendances (total)             0-24   
132195  GP attendances (total)             0-24   
132196  GP attendances (total)             0-24   

                                         MeasureName  MeasureValue Flag  \
132192  Percentage of people who had the service (%)       83.76%   NaN   
132193                       Services per 100 people        435.77  NaN   
132194              Medicare benefits per 100 people     19,788.53  NaN   
132195                               No. of patients     6,578,439  NaN   
132196                               No. of services    34,223,467  NaN   

       FlagExplanation  
132192             NaN  
132193             NaN  
132194             NaN  
132195             NaN  
132196             NaN

In [76]:
print(df_mbs_2021_sa3.index.name)
df_mbs_2021_sa3.shape

None


(231712, 14)

In [77]:
df_mbs_2021_sa3.isna().sum()

Worksheet                  0
Year                       0
StateTerritory          4032
GeographicUnit             0
GeographicCode             0
GeographicAreaName         0
GeographicGroup            0
ServiceLevel               0
Service                    0
DemographicGroup           0
MeasureName                0
MeasureValue               0
Flag                  215105
FlagExplanation       215105
dtype: int64

In [78]:
df_mbs_2021_sa3[df_mbs_2021_sa3['StateTerritory'].isna()]

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
132864       SA3  2021-22            NaN            SA3         004-01   
132865       SA3  2021-22            NaN            SA3         004-01   
132866       SA3  2021-22            NaN            SA3         004-01   
132867       SA3  2021-22            NaN            SA3         004-01   
132868       SA3  2021-22            NaN            SA3         004-01   
...          ...      ...            ...            ...            ...   
136891       SA3  2021-22            NaN            SA3         004-06   
136892       SA3  2021-22            NaN            SA3         004-06   
136893       SA3  2021-22            NaN            SA3         004-06   
136894       SA3  2021-22            NaN            SA3         004-06   
136895       SA3  2021-22            NaN            SA3         004-06   

                GeographicAreaName             GeographicGroup ServiceLevel  \
132864   Major cities - higher SES   Major cities - higher SES      Level 1   
132865   Major cities - higher SES   Major cities - higher SES      Level 1   
132866   Major cities - higher SES   Major cities - higher SES      Level 1   
132867   Major cities - higher SES   Major cities - higher SES      Level 1   
132868   Major cities - higher SES   Major cities - higher SES      Level 1   
...                            ...                         ...          ...   
136891  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   
136892  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   
136893  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   
136894  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   
136895  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   

                       Service DemographicGroup  \
132864  GP attendances (total)             0-24   
132865  GP attendances (total)             0-24   
132866  GP attendances (total)             0-24   
132867  GP attendances (total)             0-24   
132868  GP attendances (total)             0-24   
...                        ...              ...   
136891        Speech Pathology      All persons   
136892        Speech Pathology      All persons   
136893        Speech Pathology      All persons   
136894        Speech Pathology      All persons   
136895        Speech Pathology      All persons   

                                         MeasureName      MeasureValue Flag  \
132864  Percentage of people who had the service (%)           81.88%   NaN   
132865                       Services per 100 people            434.05  NaN   
132866              Medicare benefits per 100 people         19,555.73  NaN   
132867                               No. of patients           979,505  NaN   
132868                               No. of services         5,192,202  NaN   
...                                              ...               ...  ...   
136891                               No. of patients               181  NaN   
136892                               No. of services               460  NaN   
136893                  Total Medicare benefits paid           $26,965  NaN   
136894                           Total provider fees           $50,679  NaN   
136895                 Estimated resident population            535932  NaN   

       FlagExplanation  
132864             NaN  
132865             NaN  
132866             NaN  
132867             NaN  
132868             NaN  
...                ...  
136891             NaN  
136892             NaN  
136893             NaN  
136894             NaN  
136895             NaN  

[4032 rows x 14 columns]

State Terriory had 4,032 blanls because SA3 Group aggregages are not for a specific state or terriorty. They are showing aggregrates of each SA3 group by demographics and service.

SA3 groups:
004-01 = Major cities - higher SES
004-02 = Major cities - medium SES
004-03 = Major cities - lower SES
004-04 = Inner regional
004-05 = Outer regional
004-06 = Remote (incl. very remote)

In [79]:
# to ensure pivot table does not skip the blank stateterritory, filling in the blank cells.
df_mbs_2021_sa3_fill_na = df_mbs_2021_sa3.copy()
df_mbs_2021_sa3_fill_na['StateTerritory'].fillna('National-SA3Group', inplace=True)

In [80]:
print(df_mbs_2021_sa3_fill_na.isna().sum())
df_mbs_2021_sa3_fill_na.shape

Worksheet                  0
Year                       0
StateTerritory             0
GeographicUnit             0
GeographicCode             0
GeographicAreaName         0
GeographicGroup            0
ServiceLevel               0
Service                    0
DemographicGroup           0
MeasureName                0
MeasureValue               0
Flag                  215105
FlagExplanation       215105
dtype: int64


(231712, 14)

In [81]:
# Use pivot_table to reshape the DataFrame
df_mbs_2021_sa3_pivot = df_mbs_2021_sa3_fill_na.pivot_table(index=['Worksheet', 'Year', 'StateTerritory', 'GeographicUnit', 'GeographicCode', 
                                                           'GeographicAreaName', 'GeographicGroup', 'ServiceLevel', 'Service', 'DemographicGroup'],
                            columns='MeasureName', 
                            values='MeasureValue',
                            aggfunc='first')  # 'first' is used to pick the first value in case of duplicates

# Reset index to make the DataFrame flat
df_mbs_2021_sa3_pivot.reset_index(inplace=True)
df_mbs_2021_sa3_pivot.head(5)

MeasureName Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
0                 SA3  2021-22            ACT            SA3          80101   
1                 SA3  2021-22            ACT            SA3          80101   
2                 SA3  2021-22            ACT            SA3          80101   
3                 SA3  2021-22            ACT            SA3          80101   
4                 SA3  2021-22            ACT            SA3          80101   

MeasureName GeographicAreaName            GeographicGroup ServiceLevel  \
0                    Belconnen  Major cities - medium SES      Level 1   
1                    Belconnen  Major cities - medium SES      Level 1   
2                    Belconnen  Major cities - medium SES      Level 1   
3                    Belconnen  Major cities - medium SES      Level 1   
4                    Belconnen  Major cities - medium SES      Level 1   

MeasureName                            Service DemographicGroup  \
0            Allied Health attendances (total)             0-24   
1            Allied Health attendances (total)            25-44   
2            Allied Health attendances (total)            45-64   
3            Allied Health attendances (total)              65+   
4            Allied Health attendances (total)      All persons   

MeasureName Estimated resident population Medicare benefits per 100 people  \
0                                   33439                         4,385.95   
1                                   34495                         6,177.09   
2                                   22397                         6,283.74   
3                                   15541                         7,940.67   
4                                  105872                         5,892.81   

MeasureName No. of patients No. of services  \
0                     6,371          16,206   
1                     8,921          23,460   
2                     9,848          19,883   
3                     9,794          21,140   
4                    34,933          80,690   

MeasureName Percentage of people who had the service (%)  \
0                                                19.05%    
1                                                25.86%    
2                                                43.97%    
3                                                63.02%    
4                                                33.00%    

MeasureName Services per 100 people Total Medicare benefits paid  \
0                             48.46                   $1,466,617   
1                             68.01                   $2,130,787   
2                             88.78                   $1,407,369   
3                            136.03                   $1,234,059   
4                             76.21                   $6,238,832   

MeasureName Total provider fees  
0                    $2,259,631  
1                    $3,248,640  
2                    $1,933,615  
3                    $1,594,820  
4                    $9,036,706

In [82]:
df_mbs_2021_sa3_pivot.shape

(28964, 18)

In [83]:
df_mbs_2021_sa3_pivot.isna().sum()

MeasureName
Worksheet                                       0
Year                                            0
StateTerritory                                  0
GeographicUnit                                  0
GeographicCode                                  0
GeographicAreaName                              0
GeographicGroup                                 0
ServiceLevel                                    0
Service                                         0
DemographicGroup                                0
Estimated resident population                   0
Medicare benefits per 100 people                0
No. of patients                                 0
No. of services                                 0
Percentage of people who had the service (%)    0
Services per 100 people                         0
Total Medicare benefits paid                    0
Total provider fees                             0
dtype: int64

In [84]:
#removed the column index 'MeasureValue' to simply the table
df_mbs_2021_sa3_pivot.columns.name = None
df_mbs_2021_sa3_pivot.head()

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
0       SA3  2021-22            ACT            SA3          80101   
1       SA3  2021-22            ACT            SA3          80101   
2       SA3  2021-22            ACT            SA3          80101   
3       SA3  2021-22            ACT            SA3          80101   
4       SA3  2021-22            ACT            SA3          80101   

  GeographicAreaName            GeographicGroup ServiceLevel  \
0          Belconnen  Major cities - medium SES      Level 1   
1          Belconnen  Major cities - medium SES      Level 1   
2          Belconnen  Major cities - medium SES      Level 1   
3          Belconnen  Major cities - medium SES      Level 1   
4          Belconnen  Major cities - medium SES      Level 1   

                             Service DemographicGroup  \
0  Allied Health attendances (total)             0-24   
1  Allied Health attendances (total)            25-44   
2  Allied Health attendances (total)            45-64   
3  Allied Health attendances (total)              65+   
4  Allied Health attendances (total)      All persons   

  Estimated resident population Medicare benefits per 100 people  \
0                         33439                         4,385.95   
1                         34495                         6,177.09   
2                         22397                         6,283.74   
3                         15541                         7,940.67   
4                        105872                         5,892.81   

  No. of patients No. of services  \
0           6,371          16,206   
1           8,921          23,460   
2           9,848          19,883   
3           9,794          21,140   
4          34,933          80,690   

  Percentage of people who had the service (%) Services per 100 people  \
0                                      19.05%                    48.46   
1                                      25.86%                    68.01   
2                                      43.97%                    88.78   
3                                      63.02%                   136.03   
4                                      33.00%                    76.21   

  Total Medicare benefits paid Total provider fees  
0                   $1,466,617          $2,259,631  
1                   $2,130,787          $3,248,640  
2                   $1,407,369          $1,933,615  
3                   $1,234,059          $1,594,820  
4                   $6,238,832          $9,036,706

In [85]:
# check number of rows and columns as expected
df_mbs_2021_sa3_pivot.shape

(28964, 18)

In [86]:
# counting distinct values. Checking for 'National-SA3Group'
df_mbs_2021_sa3_pivot['StateTerritory'].value_counts()

NSW                  7623
Qld                  6887
Vic                  5543
WA                   2847
SA                   2350
Tas                  1257
ACT                   840
NT                    750
National-SA3Group     504
Other Territories     279
National               84
Name: StateTerritory, dtype: int64

In [87]:
df_mbs_2021_sa3_pivot.rename(columns={'Medicare benefits per 100 people': 'Medicare benefits per 100 people ($)',
                                       'Total Medicare benefits paid':'Total Medicare benefits paid ($)', 'Total provider fees':'Total provider fees ($)'}, inplace=True)
df_mbs_2021_sa3_pivot.columns

Index(['Worksheet', 'Year', 'StateTerritory', 'GeographicUnit',
       'GeographicCode', 'GeographicAreaName', 'GeographicGroup',
       'ServiceLevel', 'Service', 'DemographicGroup',
       'Estimated resident population', 'Medicare benefits per 100 people ($)',
       'No. of patients', 'No. of services',
       'Percentage of people who had the service (%)',
       'Services per 100 people', 'Total Medicare benefits paid ($)',
       'Total provider fees ($)'],
      dtype='object')

In [88]:
df_mbs_2021_sa3_pivot.drop('Worksheet', axis=1,inplace=True)

In [89]:
df_mbs_2021_sa3_pivot.columns

Index(['Year', 'StateTerritory', 'GeographicUnit', 'GeographicCode',
       'GeographicAreaName', 'GeographicGroup', 'ServiceLevel', 'Service',
       'DemographicGroup', 'Estimated resident population',
       'Medicare benefits per 100 people ($)', 'No. of patients',
       'No. of services', 'Percentage of people who had the service (%)',
       'Services per 100 people', 'Total Medicare benefits paid ($)',
       'Total provider fees ($)'],
      dtype='object')

In [90]:
df_mbs_2021_sa3_pivot.to_csv(os.path.join(path, 'clean_datasets/mbs_data/2021-22_phc_mbs.csv'))

## 2.MBS 2019-2021 Data Cleaning & Tranformation

In [91]:
# import the mbs data files and assign to a dataframe

#import required columns
filter_cols = ["Worksheet","Year","StateTerritory","GeographicUnit","GeographicCode","GeographicAreaName","GeographicGroup", "ServiceLevel", "Service", "DemographicGroup","MeasureName","MeasureValue","Flag","FlagExplanation"]

df_mbs_201921 = pd.read_csv(os.path.join(path, 'original_datasets/mbs_data/phc-mbs-2019-2021/AIHW_MBS_1920_2021_CSV.csv'), encoding='ISO-8859-1', usecols=filter_cols, index_col=None)
df_mbs_201921.head(3)

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
0       PHN  2019-20       National            NAT         001NAT   
1       PHN  2019-20       National            NAT         001NAT   
2       PHN  2019-20       National            NAT         001NAT   

  GeographicAreaName GeographicGroup ServiceLevel                 Service  \
0           National        National      Level 1  GP attendances (total)   
1           National        National      Level 1  GP attendances (total)   
2           National        National      Level 1  GP attendances (total)   

  DemographicGroup                                   MeasureName MeasureValue  \
0             0-14  Percentage of people who had the service (%)        85.57   
1            15-24  Percentage of people who had the service (%)        75.21   
2            25-44  Percentage of people who had the service (%)        82.54   

  Flag FlagExplanation  
0  NaN             NaN  
1  NaN             NaN  
2  NaN             NaN

In [92]:
df_mbs_201921.shape

(628437, 14)

In [93]:
# check the worksheet names to find SA3 reference
df_mbs_201921['Worksheet'].unique()

array(['PHN', 'SA3', 'PHN - Age standardised', 'PHN - GP aged care'],
      dtype=object)

In [94]:
# check for expected rows for SA3
df_mbs_201921[df_mbs_201921['Worksheet']== 'SA3'].count()

Worksheet             400841
Year                  400841
StateTerritory        393869
GeographicUnit        400841
GeographicCode        400841
GeographicAreaName    400841
GeographicGroup       400841
ServiceLevel          400841
Service               400841
DemographicGroup      400841
MeasureName           400841
MeasureValue          400841
Flag                   27993
FlagExplanation        27993
dtype: int64

6972 values are missing for StateTerritory. 

### 2.1 Extract and Transform SA3 Worksheet Data

In [95]:
df_mbs_201921_sa3 = df_mbs_201921[df_mbs_201921['Worksheet']== 'SA3'].copy()
df_mbs_201921_sa3

Worksheet     Year     StateTerritory GeographicUnit GeographicCode  \
113526       SA3  2019-20           National            NAT         001NAT   
113527       SA3  2019-20           National            NAT         001NAT   
113528       SA3  2019-20           National            NAT         001NAT   
113529       SA3  2019-20           National            NAT         001NAT   
113530       SA3  2019-20           National            NAT         001NAT   
...          ...      ...                ...            ...            ...   
628160       SA3  2020-21  Other Territories            SA3          90104   
628161       SA3  2020-21  Other Territories            SA3          90104   
628162       SA3  2020-21  Other Territories            SA3          90104   
628163       SA3  2020-21  Other Territories            SA3          90104   
628164       SA3  2020-21  Other Territories            SA3          90104   

       GeographicAreaName             GeographicGroup ServiceLevel  \
113526           National                    National      Level 1   
113527           National                    National      Level 1   
113528           National                    National      Level 1   
113529           National                    National      Level 1   
113530           National                    National      Level 1   
...                   ...                         ...          ...   
628160     Norfolk Island  Remote (incl. very remote)      Level 3   
628161     Norfolk Island  Remote (incl. very remote)      Level 3   
628162     Norfolk Island  Remote (incl. very remote)      Level 3   
628163     Norfolk Island  Remote (incl. very remote)      Level 3   
628164     Norfolk Island  Remote (incl. very remote)      Level 3   

                                        Service DemographicGroup  \
113526                   GP attendances (total)             0-24   
113527                   GP attendances (total)            25-44   
113528                   GP attendances (total)            45-64   
113529                   GP attendances (total)              65+   
113530                   GP attendances (total)      All persons   
...                                         ...              ...   
628160                            Physiotherapy      All persons   
628161                                 Podiatry      All persons   
628162  Practice Nurse/Aboriginal Health Worker      All persons   
628163                               Psychiatry      All persons   
628164                         Speech Pathology      All persons   

                                         MeasureName MeasureValue  Flag  \
113526  Percentage of people who had the service (%)        81.36   NaN   
113527  Percentage of people who had the service (%)        82.54   NaN   
113528  Percentage of people who had the service (%)        91.83   NaN   
113529  Percentage of people who had the service (%)          100     #   
113530  Percentage of people who had the service (%)        87.44   NaN   
...                                              ...          ...   ...   
628160                       Total provider fees ($)         n.p.  n.p.   
628161                       Total provider fees ($)         n.p.  n.p.   
628162                       Total provider fees ($)         n.p.  n.p.   
628163                       Total provider fees ($)         n.p.  n.p.   
628164                       Total provider fees ($)         n.p.  n.p.   

                                          FlagExplanation  
113526                                                NaN  
113527                                                NaN  
113528                                                NaN  
113529  The number of Medicare patients is higher than...  
113530                                                NaN  
...                                                   ...  
628160                                      Not published  
628161                         

In [96]:
print(df_mbs_201921_sa3.index.name)
df_mbs_201921_sa3.shape

None


(400841, 14)

In [97]:
df_mbs_201921_sa3.isna().sum()

Worksheet                  0
Year                       0
StateTerritory          6972
GeographicUnit             0
GeographicCode             0
GeographicAreaName         0
GeographicGroup            0
ServiceLevel               0
Service                    0
DemographicGroup           0
MeasureName                0
MeasureValue               0
Flag                  372848
FlagExplanation       372848
dtype: int64

In [98]:
df_mbs_201921_sa3[df_mbs_201921_sa3['StateTerritory'].isna()]

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
113609       SA3  2019-20            NaN            SA3         004-01   
113610       SA3  2019-20            NaN            SA3         004-01   
113611       SA3  2019-20            NaN            SA3         004-01   
113612       SA3  2019-20            NaN            SA3         004-01   
113613       SA3  2019-20            NaN            SA3         004-01   
...          ...      ...            ...            ...            ...   
600102       SA3  2020-21            NaN            SA3         004-06   
600103       SA3  2020-21            NaN            SA3         004-06   
600104       SA3  2020-21            NaN            SA3         004-06   
600105       SA3  2020-21            NaN            SA3         004-06   
600106       SA3  2020-21            NaN            SA3         004-06   

                GeographicAreaName             GeographicGroup ServiceLevel  \
113609   Major cities - higher SES   Major cities - higher SES      Level 1   
113610   Major cities - higher SES   Major cities - higher SES      Level 1   
113611   Major cities - higher SES   Major cities - higher SES      Level 1   
113612   Major cities - higher SES   Major cities - higher SES      Level 1   
113613   Major cities - higher SES   Major cities - higher SES      Level 1   
...                            ...                         ...          ...   
600102  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   
600103  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   
600104  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   
600105  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   
600106  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   

                                        Service DemographicGroup  \
113609                   GP attendances (total)             0-24   
113610                   GP attendances (total)            25-44   
113611                   GP attendances (total)            45-64   
113612                   GP attendances (total)              65+   
113613                   GP attendances (total)      All persons   
...                                         ...              ...   
600102                            Physiotherapy      All persons   
600103                                 Podiatry      All persons   
600104  Practice Nurse/Aboriginal Health Worker      All persons   
600105                               Psychiatry      All persons   
600106                         Speech Pathology      All persons   

                                         MeasureName MeasureValue Flag  \
113609  Percentage of people who had the service (%)        74.41  NaN   
113610  Percentage of people who had the service (%)        74.46  NaN   
113611  Percentage of people who had the service (%)        91.81  NaN   
113612  Percentage of people who had the service (%)          100    #   
113613  Percentage of people who had the service (%)        82.43  NaN   
...                                              ...          ...  ...   
600102                       Total provider fees ($)      1341737  NaN   
600103                       Total provider fees ($)      1485522  NaN   
600104                       Total provider fees ($)      3072378  NaN   
600105                       Total provider fees ($)      2939509  NaN   
600106                       Total provider fees ($)        58552  NaN   

                                          FlagExplanation  
113609                                                NaN  
113610                                                NaN  
113611                                                NaN  
113612  The number of Medicare patients is higher than...  
113613                                                NaN  
...                                                   ...  
600102                                     

State Terriory had 6972 blanks because SA3 Group aggregages are not for a specific state or terriorty. They are showing aggregrates of each SA3 group by demographics and service.

SA3 groups:
004-01 = Major cities - higher SES
004-02 = Major cities - medium SES
004-03 = Major cities - lower SES
004-04 = Inner regional
004-05 = Outer regional
004-06 = Remote (incl. very remote)

In [99]:
# to ensure pivot table does not skip the blank stateterritory, filling in the blank cells.
df_mbs_201921_sa3_fill_na = df_mbs_201921_sa3.copy()
df_mbs_201921_sa3_fill_na['StateTerritory'].fillna('National-SA3Group', inplace=True)

In [100]:
print(df_mbs_201921_sa3_fill_na.isna().sum())
df_mbs_201921_sa3_fill_na.shape

Worksheet                  0
Year                       0
StateTerritory             0
GeographicUnit             0
GeographicCode             0
GeographicAreaName         0
GeographicGroup            0
ServiceLevel               0
Service                    0
DemographicGroup           0
MeasureName                0
MeasureValue               0
Flag                  372848
FlagExplanation       372848
dtype: int64


(400841, 14)

In [101]:
df_mbs_201921_sa3_fill_na[['StateTerritory','GeographicGroup']].value_counts()

StateTerritory     GeographicGroup           
Qld                Major cities - medium SES     34860
NSW                Inner regional                30212
Vic                Major cities - medium SES     25564
                   Inner regional                24402
NSW                Major cities - higher SES     22078
                   Major cities - lower SES      20916
Qld                Inner regional                20916
NSW                Major cities - medium SES     18578
WA                 Major cities - medium SES     16268
Qld                Outer regional                13937
                   Major cities - higher SES     11620
Vic                Major cities - lower SES      10458
                   Major cities - higher SES     10458
Qld                Major cities - lower SES      10458
SA                 Major cities - medium SES     10451
Tas                Inner regional                 9289
NSW                Outer regional                 9282
SA                 

In [102]:
#grouping columns to check how many unique rows expected from pivot table. Expected 57263 rows
count_unique_rows = df_mbs_201921_sa3_fill_na.groupby(by=['Worksheet', 'Year', 'StateTerritory', 'GeographicUnit', 'GeographicCode', 'GeographicAreaName', 'GeographicGroup',
                               'ServiceLevel', 'Service', 'DemographicGroup'], as_index=False).agg({'MeasureValue' : 'count'})
count_unique_rows.shape

(57263, 11)

In [103]:
# Use pivot_table to reshape the DataFrame
df_mbs_201921_sa3_pivot = df_mbs_201921_sa3_fill_na.pivot_table(index=['Worksheet', 'Year', 'StateTerritory', 'GeographicUnit', 'GeographicCode', 
                                                           'GeographicAreaName', 'GeographicGroup', 'ServiceLevel', 'Service', 'DemographicGroup'],
                            columns='MeasureName', 
                            values='MeasureValue',
                            aggfunc='first')  # 'first' is used to pick the first value in case of duplicates

# Reset index to make the DataFrame flat
df_mbs_201921_sa3_pivot.reset_index(inplace=True)
df_mbs_201921_sa3_pivot.head(5)

MeasureName Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
0                 SA3  2019-20            ACT            SA3          80101   
1                 SA3  2019-20            ACT            SA3          80101   
2                 SA3  2019-20            ACT            SA3          80101   
3                 SA3  2019-20            ACT            SA3          80101   
4                 SA3  2019-20            ACT            SA3          80101   

MeasureName GeographicAreaName            GeographicGroup ServiceLevel  \
0                    Belconnen  Major cities - medium SES      Level 1   
1                    Belconnen  Major cities - medium SES      Level 1   
2                    Belconnen  Major cities - medium SES      Level 1   
3                    Belconnen  Major cities - medium SES      Level 1   
4                    Belconnen  Major cities - medium SES      Level 1   

MeasureName                            Service DemographicGroup  \
0            Allied Health attendances (total)             0-24   
1            Allied Health attendances (total)            25-44   
2            Allied Health attendances (total)            45-64   
3            Allied Health attendances (total)              65+   
4            Allied Health attendances (total)      All persons   

MeasureName Medicare benefits per 100 people ($) No. of patients  \
0                                        3275.33            6314   
1                                        5036.77            8464   
2                                        5641.12            9429   
3                                        7827.74            8973   
4                                        4986.75           33180   

MeasureName No. of services Percentage of people who had the service (%)  \
0                     14134                                        18.89   
1                     19537                                        27.12   
2                     18660                                        43.35   
3                     19925                                        61.57   
4                     72256                                        32.87   

MeasureName Services per 100 people Total Medicare benefits paid ($)  \
0                             42.29                          1094649   
1                              62.6                          1571825   
2                              85.8                          1226832   
3                            136.71                          1140815   
4                             71.58                          5034121   

MeasureName Total provider fees ($)  
0                           1612037  
1                           2307710  
2                           1605423  
3                           1421204  
4                           6946375

In [104]:
# count of rows, matches grouped dataframe. 57,363
df_mbs_201921_sa3_pivot.count()

MeasureName
Worksheet                                       57263
Year                                            57263
StateTerritory                                  57263
GeographicUnit                                  57263
GeographicCode                                  57263
GeographicAreaName                              57263
GeographicGroup                                 57263
ServiceLevel                                    57263
Service                                         57263
DemographicGroup                                57263
Medicare benefits per 100 people ($)            57263
No. of patients                                 57263
No. of services                                 57263
Percentage of people who had the service (%)    57263
Services per 100 people                         57263
Total Medicare benefits paid ($)                57263
Total provider fees ($)                         57263
dtype: int64

In [105]:
#removed the column index 'MeasureValue' to simply the table
df_mbs_201921_sa3_pivot.columns.name = None
df_mbs_201921_sa3_pivot.head()

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
0       SA3  2019-20            ACT            SA3          80101   
1       SA3  2019-20            ACT            SA3          80101   
2       SA3  2019-20            ACT            SA3          80101   
3       SA3  2019-20            ACT            SA3          80101   
4       SA3  2019-20            ACT            SA3          80101   

  GeographicAreaName            GeographicGroup ServiceLevel  \
0          Belconnen  Major cities - medium SES      Level 1   
1          Belconnen  Major cities - medium SES      Level 1   
2          Belconnen  Major cities - medium SES      Level 1   
3          Belconnen  Major cities - medium SES      Level 1   
4          Belconnen  Major cities - medium SES      Level 1   

                             Service DemographicGroup  \
0  Allied Health attendances (total)             0-24   
1  Allied Health attendances (total)            25-44   
2  Allied Health attendances (total)            45-64   
3  Allied Health attendances (total)              65+   
4  Allied Health attendances (total)      All persons   

  Medicare benefits per 100 people ($) No. of patients No. of services  \
0                              3275.33            6314           14134   
1                              5036.77            8464           19537   
2                              5641.12            9429           18660   
3                              7827.74            8973           19925   
4                              4986.75           33180           72256   

  Percentage of people who had the service (%) Services per 100 people  \
0                                        18.89                   42.29   
1                                        27.12                    62.6   
2                                        43.35                    85.8   
3                                        61.57                  136.71   
4                                        32.87                   71.58   

  Total Medicare benefits paid ($) Total provider fees ($)  
0                          1094649                 1612037  
1                          1571825                 2307710  
2                          1226832                 1605423  
3                          1140815                 1421204  
4                          5034121                 6946375

In [106]:
df_mbs_201921_sa3_pivot.drop('Worksheet', inplace=True, axis=1)

In [107]:
df_mbs_201921_sa3_pivot.to_csv(os.path.join(path, 'clean_datasets/mbs_data/2019-21_phc_mbs.csv'))

## 3.MBS 2013-2019 Data Cleaning & Tranformation

In [108]:
# import the mbs data files and assign to a dataframe

#import required columns
filter_cols = ["Worksheet","Year","StateTerritory","GeographicUnit","GeographicCode","GeographicAreaName","GeographicGroup", "ServiceLevel", "Service", "DemographicGroup","MeasureName","MeasureValue","Flag","FlagExplanation"]

df_mbs_201319 = pd.read_csv(os.path.join(path, 'original_datasets/mbs_data/phc-mbs-2013-2019/AIHW_MBS_1314_1819_CSV.csv'), encoding='ISO-8859-1', usecols=filter_cols, index_col=None)
df_mbs_201319.head(3)

/var/folders/6w/mdyngzzd2_z7brmk404qky_m0000gs/T/ipykernel_70532/3792998903.py:6: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  df_mbs_201319 = pd.read_csv(os.path.join(path, 'original_datasets/mbs_data/phc-mbs-2013-2019/AIHW_MBS_1314_1819_CSV.csv'), encoding='ISO-8859-1', usecols=filter_cols, index_col=None)


Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
0       PHN  2018-19            NAT            NAT         001NAT   
1       PHN  2018-19            NAT            NAT         001NAT   
2       PHN  2018-19            NAT            NAT         001NAT   

  GeographicAreaName GeographicGroup ServiceLevel                 Service  \
0           National        National      Level 1  GP attendances (total)   
1           National        National      Level 1  GP attendances (total)   
2           National        National      Level 1  GP attendances (total)   

  DemographicGroup                                   MeasureName  \
0             0-14  Percentage of people who had the service (%)   
1             0-14                       Services per 100 people   
2             0-14          Medicare benefits per 100 people ($)   

   MeasureValue Flag FlagExplanation  
0         86.05  NaN             NaN  
1        434.16  NaN             NaN  
2      19920.00  NaN             NaN

In [109]:
df_mbs_201319.shape

(1850655, 14)

In [110]:
# check the worksheet names to find SA3 reference
df_mbs_201319['Worksheet'].unique()

array(['PHN', 'SA3', 'PHN - Age standardised', 'PHN - GP aged care',
       'Private health insurance'], dtype=object)

#### 3.1 Extract and Transform SA3 Worksheet Data

In [111]:
df_mbs_201319_sa3 = df_mbs_201319[df_mbs_201319['Worksheet']== 'SA3'].copy()
df_mbs_201319_sa3.shape

(1206156, 14)

In [112]:
df_mbs_201319_sa3.isnull().sum()

Worksheet                   0
Year                        0
StateTerritory          20916
GeographicUnit              0
GeographicCode              0
GeographicAreaName          0
GeographicGroup             0
ServiceLevel                0
Service                     0
DemographicGroup            0
MeasureName                 0
MeasureValue           128534
Flag                  1076623
FlagExplanation       1076623
dtype: int64

20,916 values are missing for StateTerritory. 
128534 values are missing for MeasureValue.

3.1.1 Fill in missing values

In [113]:
df_mbs_201319_sa3[df_mbs_201319_sa3['StateTerritory'].isnull()]

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
113526        SA3  2018-19            NaN            SA3         004-01   
113527        SA3  2018-19            NaN            SA3         004-01   
113528        SA3  2018-19            NaN            SA3         004-01   
113529        SA3  2018-19            NaN            SA3         004-01   
113530        SA3  2018-19            NaN            SA3         004-01   
...           ...      ...            ...            ...            ...   
1827194       SA3  2013-14            NaN            SA3         004-06   
1827195       SA3  2013-14            NaN            SA3         004-06   
1827196       SA3  2013-14            NaN            SA3         004-06   
1827197       SA3  2013-14            NaN            SA3         004-06   
1827198       SA3  2013-14            NaN            SA3         004-06   

                 GeographicAreaName             GeographicGroup ServiceLevel  \
113526    Major cities - higher SES   Major cities - higher SES      Level 1   
113527    Major cities - higher SES   Major cities - higher SES      Level 1   
113528    Major cities - higher SES   Major cities - higher SES      Level 1   
113529    Major cities - higher SES   Major cities - higher SES      Level 1   
113530    Major cities - higher SES   Major cities - higher SES      Level 1   
...                             ...                         ...          ...   
1827194  Remote (incl. very remote)  Remote (incl. very remote)      Level 1   
1827195  Remote (incl. very remote)  Remote (incl. very remote)      Level 1   
1827196  Remote (incl. very remote)  Remote (incl. very remote)      Level 1   
1827197  Remote (incl. very remote)  Remote (incl. very remote)      Level 1   
1827198  Remote (incl. very remote)  Remote (incl. very remote)      Level 1   

                                Service DemographicGroup  \
113526           GP attendances (total)             0-24   
113527           GP attendances (total)             0-24   
113528           GP attendances (total)             0-24   
113529           GP attendances (total)             0-24   
113530           GP attendances (total)             0-24   
...                                 ...              ...   
1827194  Specialist attendances (total)            Males   
1827195  Specialist attendances (total)            Males   
1827196  Specialist attendances (total)            Males   
1827197  Specialist attendances (total)            Males   
1827198  Specialist attendances (total)            Males   

                                          MeasureName  MeasureValue Flag  \
113526   Percentage of people who had the service (%)         74.21  NaN   
113527                        Services per 100 people        368.97  NaN   
113528           Medicare benefits per 100 people ($)      17465.00  NaN   
113529                                No. of patients     934016.00  NaN   
113530                                No. of services    4643823.00  NaN   
...                                               ...           ...  ...   
1827194          Medicare benefits per 100 people ($)       2459.00  NaN   
1827195                               No. of patients      43360.00  NaN   
1827196                               No. of services      93787.00  NaN   
1827197              Total Medicare benefits paid ($)    7311731.00  NaN   
1827198                       Total provider fees ($)    9871635.00  NaN   

        FlagExplanation  
113526              NaN  
113527              NaN  
113528              NaN  
113529              NaN  
113530              NaN  
...                 ...  
1827194             NaN  
1827195             NaN  
1827196             NaN  
1827197             NaN  
1827198             NaN  

[20916 rows x 14 columns]

In [114]:
# to ensure pivot table does not skip the blank stateterritory, filling in the blank cells.
df_mbs_201319_sa3_fill_na = df_mbs_201319_sa3.copy()
df_mbs_201319_sa3_fill_na['StateTerritory'].fillna('National-SA3Group', inplace=True)

In [115]:
# checking for nulls in StateTerritory
print(df_mbs_201319_sa3_fill_na.isna().sum())
df_mbs_201319_sa3_fill_na.shape

Worksheet                   0
Year                        0
StateTerritory              0
GeographicUnit              0
GeographicCode              0
GeographicAreaName          0
GeographicGroup             0
ServiceLevel                0
Service                     0
DemographicGroup            0
MeasureName                 0
MeasureValue           128534
Flag                  1076623
FlagExplanation       1076623
dtype: int64


(1206156, 14)

In [116]:
# Fill in missing value so the lines are not removed by the pivot. Assigning value 'n.p' for 'Not Published'
df_mbs_201319_sa3_fill_na['MeasureValue'].fillna('n.p', inplace=True)

In [117]:
#checking for nulls in MeasureValue column
df_mbs_201319_sa3_fill_na.isna().sum()

Worksheet                   0
Year                        0
StateTerritory              0
GeographicUnit              0
GeographicCode              0
GeographicAreaName          0
GeographicGroup             0
ServiceLevel                0
Service                     0
DemographicGroup            0
MeasureName                 0
MeasureValue                0
Flag                  1076623
FlagExplanation       1076623
dtype: int64

In [118]:
#checking specific data for investigation (step can be ignored)
df_mbs_201319_sa3_fill_na[(df_mbs_201319_sa3_fill_na['Year'] == '2015-16') & (df_mbs_201319_sa3_fill_na['GeographicCode']=='30204') & 
                        (df_mbs_201319_sa3_fill_na['Service']=='GP Mental Health') & (df_mbs_201319_sa3_fill_na['DemographicGroup']== 'All persons')]

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
1343485       SA3  2015-16            Qld            SA3          30204   
1343486       SA3  2015-16            Qld            SA3          30204   
1343487       SA3  2015-16            Qld            SA3          30204   

        GeographicAreaName            GeographicGroup ServiceLevel  \
1343485           Sandgate  Major cities - medium SES      Level 3   
1343486           Sandgate  Major cities - medium SES      Level 3   
1343487           Sandgate  Major cities - medium SES      Level 3   

                  Service DemographicGroup  \
1343485  GP Mental Health      All persons   
1343486  GP Mental Health      All persons   
1343487  GP Mental Health      All persons   

                                          MeasureName MeasureValue Flag  \
1343485  Percentage of people who had the service (%)          7.3  NaN   
1343486                       Services per 100 people        13.31  NaN   
1343487          Medicare benefits per 100 people ($)       1081.0  NaN   

        FlagExplanation  
1343485             NaN  
1343486             NaN  
1343487             NaN

In [119]:
#checking specific data for investigation (step can be ignored)
df_mbs_201319_sa3_fill_na[(df_mbs_201319_sa3_fill_na['MeasureName'] == 'No. of services') 
                          & (df_mbs_201319_sa3_fill_na['MeasureValue'] == '7852')]

Empty DataFrame
Columns: [Worksheet, Year, StateTerritory, GeographicUnit, GeographicCode, GeographicAreaName, GeographicGroup, ServiceLevel, Service, DemographicGroup, MeasureName, MeasureValue, Flag, FlagExplanation]
Index: []

In [120]:
#grouping columns to check how many unique rows expected from pivot table. Expected 172323 rows
count_unique_rows_2 = df_mbs_201319_sa3_fill_na.groupby(by=['Worksheet', 'Year', 'StateTerritory', 'GeographicUnit', 'GeographicCode', 'GeographicAreaName', 'GeographicGroup',
                               'ServiceLevel', 'Service', 'DemographicGroup'], as_index=False).agg({'MeasureValue' : 'count'})
count_unique_rows_2

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
0            SA3  2013-14            ACT            SA3          80101   
1            SA3  2013-14            ACT            SA3          80101   
2            SA3  2013-14            ACT            SA3          80101   
3            SA3  2013-14            ACT            SA3          80101   
4            SA3  2013-14            ACT            SA3          80101   
...          ...      ...            ...            ...            ...   
172318       SA3  2018-19             WA            SA3          51104   
172319       SA3  2018-19             WA            SA3          51104   
172320       SA3  2018-19             WA            SA3          51104   
172321       SA3  2018-19             WA            SA3          51104   
172322       SA3  2018-19             WA            SA3          51104   

       GeographicAreaName             GeographicGroup ServiceLevel  \
0               Belconnen   Major cities - medium SES      Level 1   
1               Belconnen   Major cities - medium SES      Level 1   
2               Belconnen   Major cities - medium SES      Level 1   
3               Belconnen   Major cities - medium SES      Level 1   
4               Belconnen   Major cities - medium SES      Level 1   
...                   ...                         ...          ...   
172318           Mid West  Remote (incl. very remote)      Level 3   
172319           Mid West  Remote (incl. very remote)      Level 3   
172320           Mid West  Remote (incl. very remote)      Level 3   
172321           Mid West  Remote (incl. very remote)      Level 3   
172322           Mid West  Remote (incl. very remote)      Level 3   

                                        Service DemographicGroup  MeasureValue  
0             Allied Health attendances (total)             0-24             7  
1             Allied Health attendances (total)            25-44             7  
2             Allied Health attendances (total)            45-64             7  
3             Allied Health attendances (total)              65+             7  
4             Allied Health attendances (total)      All persons             7  
...                                         ...              ...           ...  
172318                            Physiotherapy      All persons             7  
172319                                 Podiatry      All persons             7  
172320  Practice Nurse/Aboriginal Health Worker      All persons             7  
172321                               Psychiatry      All persons             7  
172322                         Speech Pathology      All persons             7  

[172323 rows x 11 columns]

In [121]:
#creatiing unique key by combining columns.  This is to compare missing key with pivot table
count_unique_rows_2['key'] = count_unique_rows_2['Worksheet'].astype(str) + count_unique_rows_2['Year'].astype(str) + count_unique_rows_2['StateTerritory'].astype(str) + count_unique_rows_2['GeographicUnit'].astype(str) + count_unique_rows_2['GeographicCode'].astype(str) + count_unique_rows_2['GeographicAreaName'].astype(str) + count_unique_rows_2['GeographicGroup'].astype(str) + count_unique_rows_2['ServiceLevel'].astype(str) + count_unique_rows_2['Service'].astype(str) + count_unique_rows_2['DemographicGroup'].astype(str)
count_unique_rows_2['key']

0         SA32013-14ACTSA380101BelconnenMajor cities - m...
1         SA32013-14ACTSA380101BelconnenMajor cities - m...
2         SA32013-14ACTSA380101BelconnenMajor cities - m...
3         SA32013-14ACTSA380101BelconnenMajor cities - m...
4         SA32013-14ACTSA380101BelconnenMajor cities - m...
                                ...                        
172318    SA32018-19WASA351104Mid WestRemote (incl. very...
172319    SA32018-19WASA351104Mid WestRemote (incl. very...
172320    SA32018-19WASA351104Mid WestRemote (incl. very...
172321    SA32018-19WASA351104Mid WestRemote (incl. very...
172322    SA32018-19WASA351104Mid WestRemote (incl. very...
Name: key, Length: 172323, dtype: object

In [122]:
# investigating why 2 rows are generated for these 15 rows by pivotable and grouping
dups = ["SA32013-14VicSA321502MilduraOuter regionalLevel 1Allied Health attendances (total)45-64", "SA32014-15VicSA320401Upper Goulburn ValleyInner regionalLevel 3GP Mental HealthAll persons", "SA32014-15VicSA321704WarrnamboolInner regionalLevel 1Nursing and Aboriginal Health Workers (total)65+",
        "SA32015-16QldSA330204SandgateMajor cities - medium SESLevel 3GP Mental HealthAll persons",
        "SA32015-16QldSA331402North LakesMajor cities - medium SESLevel 1Nursing and Aboriginal Health Workers (total)65+",
        "SA32016-17NSWSA310106Young - YassInner regionalLevel 1Nursing and Aboriginal Health Workers (total)All persons",
        "SA32016-17NSWSA311903HurstvilleMajor cities - medium SESLevel 1Allied Health attendances (total)65+",
        "SA32016-17QldSA330402Kenmore - Brookfield - MoggillMajor cities - higher SESLevel 2Allied Health subtotal - OptometryAll persons",
        "SA32016-17QldSA331605NoosaMajor cities - medium SESLevel 3GP Mental HealthAll persons",
        "SA32017-18NSWSA310402Coffs HarbourInner regionalLevel 3GP Multidisciplinary Case ConferenceAll persons",
        "SA32017-18NSWSA312103Ku-ring-gaiMajor cities - higher SESLevel 1Nursing and Aboriginal Health Workers (total)All persons",
        "SA32017-18QldSA330601Cairns - NorthOuter regionalLevel 3Cervical Smear PIPAll persons",
        "SA32017-18QldSA331902BurnettInner regionalLevel 2Allied Health subtotal - OptometryAll persons",
        "SA32018-19NSWSA310801Great LakesInner regionalLevel 1Allied Health attendances (total)45-64",
        "SA32018-19TasSA360105Hobart InnerInner regionalLevel 1Allied Health attendances (total)45-64"]
x = count_unique_rows_2['key'].isin(['SA32013-14VicSA321502MilduraOuter regionalLevel 1Allied Health attendances (total)45-64'])
count_unique_rows_2[x]

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
25081       SA3  2013-14            Vic            SA3          21502   
25253       SA3  2013-14            Vic            SA3          21502   

      GeographicAreaName GeographicGroup ServiceLevel  \
25081            Mildura  Outer regional      Level 1   
25253            Mildura  Outer regional      Level 1   

                                 Service DemographicGroup  MeasureValue  \
25081  Allied Health attendances (total)            45-64             1   
25253  Allied Health attendances (total)            45-64             6   

                                                     key  
25081  SA32013-14VicSA321502MilduraOuter regionalLeve...  
25253  SA32013-14VicSA321502MilduraOuter regionalLeve...

##### 3.1.2 Pivot MBS data by Measure Name / Value

In [123]:
# Use pivot_table to reshape the DataFrame
df_mbs_201319_sa3_pivot = df_mbs_201319_sa3_fill_na.pivot_table(index=['Worksheet', 'Year', 'StateTerritory', 'GeographicUnit', 'GeographicCode', 
                                                           'GeographicAreaName', 'GeographicGroup', 'ServiceLevel', 'Service', 'DemographicGroup'],
                            columns='MeasureName', 
                            values='MeasureValue',
                            aggfunc='first')  # 'first' is used to pick the first value in case of duplicates

# Reset index to make the DataFrame flat
df_mbs_201319_sa3_pivot.reset_index(inplace=True)
df_mbs_201319_sa3_pivot.head(5)

MeasureName Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
0                 SA3  2013-14            ACT            SA3          80101   
1                 SA3  2013-14            ACT            SA3          80101   
2                 SA3  2013-14            ACT            SA3          80101   
3                 SA3  2013-14            ACT            SA3          80101   
4                 SA3  2013-14            ACT            SA3          80101   

MeasureName GeographicAreaName            GeographicGroup ServiceLevel  \
0                    Belconnen  Major cities - medium SES      Level 1   
1                    Belconnen  Major cities - medium SES      Level 1   
2                    Belconnen  Major cities - medium SES      Level 1   
3                    Belconnen  Major cities - medium SES      Level 1   
4                    Belconnen  Major cities - medium SES      Level 1   

MeasureName                            Service DemographicGroup  \
0            Allied Health attendances (total)             0-24   
1            Allied Health attendances (total)            25-44   
2            Allied Health attendances (total)            45-64   
3            Allied Health attendances (total)              65+   
4            Allied Health attendances (total)      All persons   

MeasureName Medicare benefits per 100 people ($) No. of patients  \
0                                         2576.0          5624.0   
1                                         4004.0          7714.0   
2                                         4672.0          8998.0   
3                                         5819.0          6397.0   
4                                         3892.0         28733.0   

MeasureName No. of services Percentage of people who had the service (%)  \
0                   10879.0                                        17.27   
1                   15870.0                                        24.75   
2                   15754.0                                        41.32   
3                   12316.0                                        55.07   
4                   54818.0                                        29.59   

MeasureName Services per 100 people Total Medicare benefits paid ($)  \
0                             33.41                         838549.0   
1                             50.93                        1247656.0   
2                             72.35                        1017264.0   
3                            106.01                         675946.0   
4                             56.45                        3779415.0   

MeasureName Total provider fees ($)  
0                         1026474.0  
1                         1600846.0  
2                         1197133.0  
3                          761837.0  
4                         4586290.0

In [124]:
# count of rows, matches grouped dataframe. 172,323
df_mbs_201319_sa3_pivot.shape

(172323, 17)

In [125]:
df_mbs_201319_sa3_pivot['key'] = df_mbs_201319_sa3_pivot['Worksheet'].astype(str) + df_mbs_201319_sa3_pivot['Year'].astype(str) + df_mbs_201319_sa3_pivot['StateTerritory'].astype(str) + df_mbs_201319_sa3_pivot['GeographicUnit'].astype(str) + df_mbs_201319_sa3_pivot['GeographicCode'].astype(str) + df_mbs_201319_sa3_pivot['GeographicAreaName'].astype(str) + df_mbs_201319_sa3_pivot['GeographicGroup'].astype(str) + df_mbs_201319_sa3_pivot['ServiceLevel'].astype(str) + df_mbs_201319_sa3_pivot['Service'].astype(str) + df_mbs_201319_sa3_pivot['DemographicGroup'].astype(str)
df_mbs_201319_sa3_pivot['key'].value_counts

<bound method IndexOpsMixin.value_counts of 0         SA32013-14ACTSA380101BelconnenMajor cities - m...
1         SA32013-14ACTSA380101BelconnenMajor cities - m...
2         SA32013-14ACTSA380101BelconnenMajor cities - m...
3         SA32013-14ACTSA380101BelconnenMajor cities - m...
4         SA32013-14ACTSA380101BelconnenMajor cities - m...
                                ...                        
172318    SA32018-19WASA351104Mid WestRemote (incl. very...
172319    SA32018-19WASA351104Mid WestRemote (incl. very...
172320    SA32018-19WASA351104Mid WestRemote (incl. very...
172321    SA32018-19WASA351104Mid WestRemote (incl. very...
172322    SA32018-19WASA351104Mid WestRemote (incl. very...
Name: key, Length: 172323, dtype: object>

In [126]:
df_mbs_201319_sa3_pivot.isna().sum()

MeasureName
Worksheet                                        0
Year                                             0
StateTerritory                                   0
GeographicUnit                                   0
GeographicCode                                   0
GeographicAreaName                               0
GeographicGroup                                  0
ServiceLevel                                     0
Service                                          0
DemographicGroup                                 0
Medicare benefits per 100 people ($)            15
No. of patients                                 15
No. of services                                 15
Percentage of people who had the service (%)    15
Services per 100 people                         15
Total Medicare benefits paid ($)                15
Total provider fees ($)                         15
key                                              0
dtype: int64

In [127]:
df_mbs_201319_sa3_pivot[df_mbs_201319_sa3_pivot.isna().any(axis=1)]

MeasureName Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
25081             SA3  2013-14            Vic            SA3          21502   
25253             SA3  2013-14            Vic            SA3          21502   
49168             SA3  2014-15            Vic            SA3          20401   
53772             SA3  2014-15            Vic            SA3          21704   
54595             SA3  2014-15            Vic            SA3          20401   
54604             SA3  2014-15            Vic            SA3          21704   
67513             SA3  2015-16            Qld            SA3          30204   
72117             SA3  2015-16            Qld            SA3          31402   
72691             SA3  2015-16            Qld            SA3          30204   
72700             SA3  2015-16            Qld            SA3          31402   
86989             SA3  2016-17            NSW            SA3          10106   
91652             SA3  2016-17            NSW            SA3          11903   
92028             SA3  2016-17            NSW            SA3          10106   
92075             SA3  2016-17            NSW            SA3          11903   
96239             SA3  2016-17            Qld            SA3          30402   
100877            SA3  2016-17            Qld            SA3          31605   
102049            SA3  2016-17            Qld            SA3          30402   
102084            SA3  2016-17            Qld            SA3          31605   
115742            SA3  2017-18            NSW            SA3          10402   
120346            SA3  2017-18            NSW            SA3          12103   
121419            SA3  2017-18            NSW            SA3          10402   
121428            SA3  2017-18            NSW            SA3          12103   
124948            SA3  2017-18            Qld            SA3          30601   
129604            SA3  2017-18            Qld            SA3          31902   
131374            SA3  2017-18            Qld            SA3          30601   
131441            SA3  2017-18            Qld            SA3          31902   
144433            SA3  2018-19            NSW            SA3          10801   
152055            SA3  2018-19            NSW            SA3          10801   
163111            SA3  2018-19            Tas            SA3          60105   
163126            SA3  2018-19            Tas            SA3          60105   

MeasureName              GeographicAreaName            GeographicGroup  \
25081                               Mildura             Outer regional   
25253                               Mildura             Outer regional   
49168                 Upper Goulburn Valley             Inner regional   
53772                           Warrnambool             Inner regional   
54595                 Upper Goulburn Valley             Inner regional   
54604                           Warrnambool             Inner regional   
67513                              Sandgate  Major cities - medium SES   
72117                           North Lakes  Major cities - medium SES   
72691                              Sandgate  Major cities - medium SES   
72700                           North Lakes  Major cities - medium SES   
86989                          Young - Yass             Inner regional   
91652                            Hurstville  Major cities - medium SES   
92028                          Young - Yass             Inner regional   
92075                            Hurstville  Major cities - medium SES   
96239        Kenmore - Brookfield - Moggill  Major cities - higher SES   
100877                                Noosa  Major cities - medium SES   
102049       Kenmore - Brookfield - Moggill  Major cities - higher SES   
102084                                Noosa  Major cities - medium SES   
115742                        Coffs Harbour             Inner regional   
120346                          Ku-ring-gai  Major cities - higher SES

In [128]:
df_mbs_201319_sa3_pivot[df_mbs_201319_sa3_pivot.duplicated(subset=['key'], keep='first')]['key'].to_clipboard()

In [129]:
df_mbs_201319_sa3_pivot[(df_mbs_201319_sa3_pivot['Year'] == '2015-16') & (df_mbs_201319_sa3_pivot['GeographicCode']=='30204') & 
                        (df_mbs_201319_sa3_pivot['Service']=='GP Mental Health') & (df_mbs_201319_sa3_pivot['DemographicGroup']== 'All persons')]

MeasureName Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
72691             SA3  2015-16            Qld            SA3          30204   

MeasureName GeographicAreaName            GeographicGroup ServiceLevel  \
72691                 Sandgate  Major cities - medium SES      Level 3   

MeasureName           Service DemographicGroup  \
72691        GP Mental Health      All persons   

MeasureName Medicare benefits per 100 people ($) No. of patients  \
72691                                     1081.0             NaN   

MeasureName No. of services Percentage of people who had the service (%)  \
72691                   NaN                                          7.3   

MeasureName Services per 100 people Total Medicare benefits paid ($)  \
72691                         13.31                              NaN   

MeasureName Total provider fees ($)  \
72691                           NaN   

MeasureName                                                key  
72691        SA32015-16QldSA330204SandgateMajor cities - me...

In [130]:
missing_keys = count_unique_rows_2[~count_unique_rows_2['key'].isin(df_mbs_201319_sa3_pivot['key'])]

In [131]:
missing_keys.shape

(0, 12)

In [132]:
df_mbs_201319_sa3_pivot.columns

Index(['Worksheet', 'Year', 'StateTerritory', 'GeographicUnit',
       'GeographicCode', 'GeographicAreaName', 'GeographicGroup',
       'ServiceLevel', 'Service', 'DemographicGroup',
       'Medicare benefits per 100 people ($)', 'No. of patients',
       'No. of services', 'Percentage of people who had the service (%)',
       'Services per 100 people', 'Total Medicare benefits paid ($)',
       'Total provider fees ($)', 'key'],
      dtype='object', name='MeasureName')

In [133]:
# Column key will be dropped. Was generated for investigation purposes
df_mbs_201319_sa3_pivot.drop(['key', 'Worksheet'], axis=1, inplace=True)
df_mbs_201319_sa3_pivot.columns

Index(['Year', 'StateTerritory', 'GeographicUnit', 'GeographicCode',
       'GeographicAreaName', 'GeographicGroup', 'ServiceLevel', 'Service',
       'DemographicGroup', 'Medicare benefits per 100 people ($)',
       'No. of patients', 'No. of services',
       'Percentage of people who had the service (%)',
       'Services per 100 people', 'Total Medicare benefits paid ($)',
       'Total provider fees ($)'],
      dtype='object', name='MeasureName')

In [134]:
df_mbs_201319_sa3_pivot.to_csv(os.path.join(path, 'clean_datasets/mbs_data/2013-19_phc_mbs.csv'))